docs/en/observability/threshold-alert.asciidoc

```diff
@@ -0,0 +1,152 @@
++
+. Set the following aggregations:
+** *Aggregation A:* Average `system.cpu.user.pct`
+** *Aggregation B:* Average `system.cpu.system.pct`
+** *Aggregation C:* Max `system.cpu.cores`.
+. Set the equation to `(A + B) / C * 100`
+. Set the threshold to `IS ABOVE 95` to alert when CPU usage is above 95%.
+*Boolean logic*:: Use conditional operators and comparison operators with you aggregations to define conditions for alerting.
++
+*Example* +
+Set an equation and threshold to trigger an alert when two fields that should have same value differ. For this example, we'll use `kubernetes.statefulset.replicas.ready` and `kubernetes.statefulset.replicas.desired`, and trigger an alert when their values differ. To do this, set the following aggregations, equation, and threshold:
++
+. Set the following aggregations:
+** *Aggregation A:* Sum `kubernetes.statefulset.replicas.ready`
+** *Aggregation B:* Sum `kubernetes.statefulset.desired`
+. Set the equation to `A == B ? 1 : 0`. If A and B are equal, the result is `1`. If they're not equal, the result is `0`.
+. Set the threshold to `IS BELOW 1` to trigger an alert when the result is `0` and the field values do not match.
+
+[discrete]
+[[custom-threshold-chart-preview]]
+== Preview chart
+
+The preview chart provides a visualization of how many entries match your configuration.
+The shaded area shows the threshold you've set.
+
+[role="screenshot"]
+image::images/custom-threshold-preview-chart.png[Custom threshold preview chart,75%]
+
+[discrete]
+[[custom-threshold-group-by]]
+== Group alerts by (optional)
+
+Set one or more *group alerts by* fields for custom threshold rules to perform a composite aggregation against the selected fields.
+When any of these groups match the selected rule conditions, an alert is triggered _per group_.
+
+When you select multiple groupings, the group name is separated by commas.
+
+For example, if you group alerts by the `host.name` and `host.architecture` fields, and there are two hosts (`Host A` and `Host B`) and two architectures (`Architecture A` and `Architecture B`), the composite aggregation forms multiple groups.
+
+If the `Host A, Architecture A` group matches the rule conditions, but the `Host B, Architecture B` group doesn't, one alert is triggered for `Host A, Architecture A`.
+
+If you select one field--for example, `host.name`--and `Host A` matches the conditions but `Host B` doesn't, one alert is triggered for `Host A`.
+If both groups matches the conditions, alerts are triggered for both groups.
+
+[IMPORTANT]
+=====
+When *group alerts by* fields are selected, but no documents contain the selected field(s) within the given time range of when the alert is triggered, you can't determine the group(s). This is relevant when the rule condition is sensitive to a certain number of documents, and that number might be `0`.
```

Comment:
This part was a bit confusing, is it about the fact that missing group setting only is applicable when a group was seen previously? So when the rule is running, Host-1 should be seen at least once in order to report it as a missing group?

Reply:
I actually pulled the **group by** info from another rule doc, so I have made the assumption that this is still relevant. If not, please let me know.

My understanding is that if a field that you're grouping by doesn't have data during the time range of your query, your rule won't trigger alerts based on that field. So if you have a field with no data during the time range, and you have a rule configured to alert when that field is or includes `0`, you might expect the rule to trigger an alert assuming that no data and `0` are equivalent. But that wouldn't happen in this case. 

I've attempted to rewrite it, let me know if that makes more sense or I can give it another shot.

Reply:
It is still not clear to me what we are trying to mention here, I think it is relevant as this rule is similar to the Metric threshold, so the same conditions are applied in both rules.

@simianhacker Is [this warning](https://observability-docs_3314.docs-preview.app.elstc.co/guide/en/observability/master/custom-threshold-alert.html#custom-threshold-group-by) related to what you shared that when we have a group by field, `no data` alert will not be triggered anymore? (This [comment](https://github.com/elastic/sdh-kibana/issues/4159#issuecomment-1750940737)) Or is this warning about something else?

Reply:
Hum... I think we are trying to indicate that when there are ZERO documents we can't see ANY information about the groups. It's important to understand, `doc_count < 0` doesn't make sense when are alerting on `NO_DATA` because `NO_DATA` takes precedence. We can only alert on `NO_DATA` for a group by rule BUT ONLY after the rule execution has "seen" the instance of the group at least once.

Reply:
Ok, I _think_ I understand, do we want to say that if you have **group alerts by** fields selected and **Alert me if groups stop reporting data** enabled, you won't be alerted if the fields in the time range have no data. You would only be alerted if there was data at one point and it stops?

Reply:
I just noticed this group by information was copied from [log threshold](https://www.elastic.co/guide/en/observability/7.17/logs-threshold-alert.html#group-by) rule, not metric threshold, so it is not as relevant as I initially thought.

I did some tests, and I am leaning toward removing this warning. This is more applicable when there is a filter involved, and it points to the difference between 0 and `No data.`

Here are some tests that I did:
1. When we filter data for a field that does not match any document, we get alerts for all the groups: (`container-name1` does not match any document)

|Alerts|Rule definition|
|---|---|
|![image](https://github.com/elastic/observability-docs/assets/12370520/3edfce07-cb0b-4ca5-856c-3641645b0c02)|![image](https://github.com/elastic/observability-docs/assets/12370520/09783a86-5675-4c96-916c-18dda9c154e5)|

2. When I stopped sending data, I got no data for the same alert group:

<img src="https://github.com/elastic/observability-docs/assets/12370520/4a83bbbe-8d33-4461-a26a-b6542a3892fc" width=400 />

3. When I selected a group by a field that does not exist in the documents (`host.type`), no alerts were generated, but since we have the preview, it was visible there that no data is available.

<img src="https://github.com/elastic/observability-docs/assets/12370520/398f7f51-d1c0-426f-a553-7a8c957062c1" width=400 />

In conclusion, let's remove this warning and add it later if we see there is confusion about it.

Reply:
Thanks for doing that testing. Makes sense to me to remove it.